cmd/zc_attr_filter_windows.go

```diff
@@ -0,0 +1,165 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"path/filepath"
+	"syscall"
+	"strings"
+)
+
+type excludeAttrFilter struct {
+	fileAttributes uint32
+	filePath       string
+}
+
+func (f *excludeAttrFilter) doesSupportThisOS() (msg string, supported bool) {
+	msg = ""
+	supported = true
+	return
+}
+
+func (f *excludeAttrFilter) doesPass(storedObject storedObject) bool {
+	fileName := filepath.Join(f.filePath, storedObject.name)
+	lpFileName, err := syscall.UTF16PtrFromString(fileName)
+
+	// If it fails to retrive the pointer from file path, let the filter pass
+	if err != nil {
+		return true
+	}
+
+	attributes, err := syscall.GetFileAttributes(lpFileName)
+
+	if err != nil {
+		return true
+	}
+
+	// if a file shares one or more attributes with the ones provided by the filter,
+	// exclude this file
+	if attributes & f.fileAttributes > 0 {
+		return false
+	}
+
+	return true
+}
+
+func buildExcludeAttrFilters(attributes []string, fullPath string) []objectFilter {
+	var fileAttributes uint32
+	filters := make([]objectFilter, 0)
+	// Available attributes (NTFS) include:
+	// R = Read-only files
+	// A = Files ready for archiving
+	// S = System files
+	// H = Hidden files
+	// C = Compressed files
+	// N = Normal files
+	// E = Encrypted files
+	// T = Temporary files
+	// O = Offline files
+	// I = Non-indexed files
+	// Reference for File Attribute Constants:
+	// https://docs.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants
+	fileAttributeMap := map[string]uint32 {
+		"R" : 1,
+		"A" : 32,
+		"S" : 4,
+		"H" : 2,
+		"C" : 2048,
+		"N" : 128,
+		"E" : 16384,
+		"T" : 256,
+		"O" : 4096,
+		"I" : 8192,
+	}
+
+	for _, attribute := range attributes {
+		fileAttributes |= fileAttributeMap[strings.ToUpper(attribute)]
+	}
+
+	// Don't append the filter if there is no attributes given
+	if fileAttributes > 0 {
+		filters = append(filters, &excludeAttrFilter{fileAttributes: fileAttributes, filePath: fullPath})
+	}
+	return filters
+}
+
+type includeAttrFilter struct {
+	fileAttributes uint32
+	filePath       string
+}
+
+func (f *includeAttrFilter) doesSupportThisOS() (msg string, supported bool) {
+	msg = ""
+	supported = true
+	return
+}
+
+func (f *includeAttrFilter) doesPass(storedObject storedObject) bool {
+	fileName := filepath.Join(f.filePath, storedObject.name)
+	lpFileName, err := syscall.UTF16PtrFromString(fileName)
+
+	// If it fails to retrive the pointer from file path, let the filter pass
+	if err != nil {
+		return true
+	}
+
+	attributes, err := syscall.GetFileAttributes(lpFileName)
+
+	// If the storedObject has invalid file attributes, let the filter pass
+	if err != nil {
+		return true
+	}
+
+	// if a file shares one or more attributes with the ones provided by the filter,
```

Comment:
Am I correct in understanding that includeAttrFilter and excludeAttrFilter are identical expect that `return true` and `return false` here are around the other way?

If so, what about changing them so that there's only one type, called something like attrFilter, like this:
```
type attrFilter struct { 
	fileAttributes uint32 
	filePath       string 
        resultIfMatch bool
}
```

And then this code here would become
```
// if a file shares one or more attributes with the ones provided by the filter, it's a match. So return the 
// value that we are supposed to return for matches, else return the opposite
	if attributes & f.fileAttributes > 0 { 
		return f.resultIfMatch
	} 

	return !f.resultIfMatch
```

That would reduce a lot of code duplication.  BTW, maybe there is some better variation of what I've suggested that would also remove the duplication, but might not be done exactly as I've suggested. That would probably be fine too. The key thing is to minimize duplication.

(Sorry if this comment seems like nitpicking.  Based on experience with the codebase over the past 6 months, we've become very conscious of the need to control, and eventually reduce, duplication of code.  So it's something we look hard at now.)

Reply:
I modified the code like you suggested and it did reduce a lot of duplication!